k2/csrc/cuda/debug.cuh

```diff
@@ -0,0 +1,169 @@
+// k2/csrc/cuda/errors.h
+
+// Copyright (c) 2020, Xiaomi Corporation ( authors: Meixu Song )
+
+// See LICENSE for clarification regarding multiple authors
+
+#ifndef K2_CSRC_CUDA_ERRORS_H_
+#define K2_CSRC_CUDA_ERRORS_H_
+
+#include <cstdio>
+
+#include "k2/csrc/cuda/arch.h"
+
+namespace k2 {
+
+// In debug mode and include this header,
+// define K2_MAKE_ERROR_CHECK to turn on error checking
+#if (!defined(NDEBUG) && !defined(K2_MAKE_ERROR_CHECK))
+  #define K2_MAKE_ERROR_CHECK
+#endif
+
+/**
+ * \brief A static assertion
+ * \param exp the compile-time boolean expression that must be true
+ * \param msg an error message if exp is false
+ *
+ * `static_assert` is supported by both of host and device.
+ */
+#ifndef K2_STATIC_ASSERT
+  #define K2_STATIC_ASSERT(exp, msg) static_assert(exp, msg)
+#endif
+
+/**
+ * \brief This is a error checking function, with context information.
+ * If K2_MAKE_ERROR_CHECK is defined and error is not cudaSuccess,
+ * the corresponding error message is printed to:
+ *   - host: stderr
+ *   - device: stdout in device
+ * along with the supplied source context.
+ * It's used to make other macros convenient.
+ *
+ * \return The CUDA error.
+ */
+__host__ __device__ __forceinline__ cudaError_t K2_CUDA_DEBUG(
+    cudaError_t error,
+    const char *filename,
+    int line,
+    bool abort = true) {
+#ifdef K2_MAKE_ERROR_CHECK
+  (void)filename;
+  (void)line;
+  if (cudaSuccess != error) {
+  #if (K2_PTX_ARCH == 0)
+    fprintf(stderr, "CUDA error ID=%d, NAME=%s, [%s, %d]: %s\n",
+            error, cudaGetErrorName(error),
+            filename, line,
+            cudaGetErrorString(error));
+    fflush(stderr);
+  #elif (K2_PTX_ARCH >= 200)
+    printf("CUDA error ID=%d, NAME=%s, [block (%d,%d,%d) thread (%d,%d,%d), %s, %d]: %s\n",
+           error, cudaGetErrorName(error),
+           blockIdx.z, blockIdx.y, blockIdx.x, threadIdx.z, threadIdx.y, threadIdx.x,
+           filename, line,
+           cudaGetErrorString(error));
+  #endif
+    if (abort) {
+      exit(EXIT_FAILURE);
+    }
+  }
+#endif
+  return error;
+}
+
+/**
+ * \brief Macro for checking cuda error.
+ * If not cudaSuccess and K2_MAKE_ERROR_CHECK, print and exit.
```

Comment:
Be more precise: "if K2_MAKE_ERROR_CHECK is defined and e is not cudaSuccess"... and clarify that it does nothing otherwise.

Reply:
Done.